src/main/java/com/datadoghq/sketch/ddsketch/DDSketch.java

```diff
@@ -7,45 +7,35 @@
- * value is 100, the computed quantile value is guaranteed to be between 99 and 101. If the expected quantile value
- * is 1000, the computed quantile value is guaranteed to be between 990 and 1010.
+ * For instance, using {@code DDSketch} with a relative accuracy guarantee set to 1%, if the expected quantile value is
+ * 100, the computed quantile value is guaranteed to be between 99 and 101. If the expected quantile value is 1000, the
+ * computed quantile value is guaranteed to be between 990 and 1010.
  * <p>
- * {@code DDSketch} works by mapping floating-point input values to bins and counting the number of values for each
- * bin. The mapping to bins is handled by {@link IndexMapping}, while the underlying structure that keeps track of
- * bin counts is {@link Store}. {@link #memoryOptimal} constructs a sketch with a logarithmic index mapping, hence low
- * memory footprint, whereas {@link #fast} and {@link #balanced} offer faster ingestion speeds at the cost of
- * larger memory footprints. The size of the sketch can be upper-bounded by using collapsing stores. For instance,
- * {@link #memoryOptimalCollapsingLowest} is the version of {@code DDSketch} described in the paper, and also
- * implemented in <a href="https://github.com/DataDog/sketches-go/">Go</a>
- * and <a href="https://github.com/DataDog/sketches-py/">Python</a>
- * . It collapses lowest bins when the maximum number of buckets is reached. For using a specific
- * {@link IndexMapping} or a specific implementation of {@link Store}, the constructor can be used
- * ({@link #DDSketch(IndexMapping, Supplier)}).
+ * {@code DDSketch} works by mapping floating-point input values to bins and counting the number of values for each bin.
+ * The mapping to bins is handled by {@link IndexMapping}, while the underlying structure that keeps track of bin counts
+ * is {@link Store}. The {@link IndexMapping} and the {@link Store} supplier are provided when constructing the sketch
+ * and can be adjusted based on use-case requirements. See {@link DDSketches} for preset sketches and some explanation
+ * about the involved tradeoffs.
  * <p>
- * The memory size of the sketch depends on the range that is covered by the input values: the larger that range, the
- * more bins are needed to keep track of the input values. As a rough estimate, if working on durations using
- * {@link #memoryOptimal} with a relative accuracy of 2%, about 2kB (275 bins) are needed to cover values between 1
- * millisecond and 1 minute, and about 6kB (802 bins) to cover values between 1 nanosecond and 1 day. The number of
- * bins that are maintained can be upper-bounded using collapsing stores (see for example
- * {@link #memoryOptimalCollapsingLowest} and {@link #memoryOptimalCollapsingHighest}).
+ * Note that negative values are inverted before being mapped to the store. That means that if you use a store that
```

Comment:
If the positive store is CollapsingLowest shouldn't we use CollapsingHighest for the negative store?

Reply:
It's hard to say without knowing more concrete use cases, but I'd tend to think that if that were the case, we would be interested in highest quantiles and we probably wouldn't want a collapsing store for positive values in the first place (which would cause inaccurate quantiles in two non-contiguous areas). Or we would want a mechanism that starts collapsing the positive-value store only when the negative-value store is fully collapsed.

I believe the idea of collapsing close to zero fits more use cases, where we would allow loosening the relative-accuracy guarantee with an (adaptive) absolute-accuracy guarantee. Said otherwise, that would be a relatively accurate sketch on a best effort basis, which would keep the absolute error as low as possible when the relative accuracy cannot be enforced. For instance, if we are plotting those quantile values (e.g., as a time-series over time), I believe that's what we want, as opposed to collapsing on the right sides (or left sides) of both stores.

What do you think?

Reply:
in the python version we're collapsing the negative store first (using highest since it's reversed) and then collapsing the positive store. so the left collapse of the positive store doesn't happen until the negative store is empty